plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar310/TreeBuilder310.java

```diff
@@ -632,76 +634,250 @@ public final SimpleNode onCloseNode(SimpleNode n, int arity) throws Exception {
+                            return new MatchClass(null, null, new String[] { rep }, new patternType[] { pattern });
+                        } else {
+                            Log.log("Expected arity to be 2 here.");
+                        }
+                    } else {
+                        Name[] names = new Name[arity];
+                        try {
+                            names[0] = (Name) popNode;
+                        } catch (Exception e) {
+                            Log.log("Expected Name. Found: " + popNode);
+                        }
+                        for (int i = 1; i < arity; i++) {
+                            popNode = stack.popNode();
+                            try {
+                                names[i] = (Name) popNode;
+                            } catch (Exception e) {
+                                Log.log("Expected Name. Found: " + popNode);
+                            }
+                        }
+                        Attribute[] attrs = new Attribute[names.length - 1];
+                        for (int i = 0; i < attrs.length; i++) {
+                            NameTok attr = new NameTok(names[i].id, names[i].ctx);
+                            attrs[i] = new Attribute(null, attr, attr.ctx);
+                        }
+                        for (int i = 0; i < attrs.length - 1; i++) {
+                            attrs[i].value = attrs[i + 1];
+                        }
+                        attrs[attrs.length - 1].value = names[names.length - 1];
+                        return attrs[0];
+                    }
+                } else if (arity == 1) {
+                    return stack.popNode();
+                } else {
+                    Log.log("Expected arity to be > 0 here.");
+                }
+                return null;
+
+            case JJTCLASS_PATTERN:
+                exprType cls = null;
+                String[] kwd_attrs = null;
+                patternType[] kwd_patterns = null;
+                if (arity > 1) {
+                    patternType[] patterns = null;
+                    patterns = new patternType[arity - 1];
+                    for (int i = 0; i < patterns.length; i++) {
+                        SimpleNode popNode = stack.popNode();
+                        try {
+                            patterns[i] = (patternType) popNode;
+                        } catch (Exception e) {
+                            Log.log("Expected pattern. Found: " + popNode);
+                        }
+                    }
+                    SimpleNode popNode = stack.peekNode();
+                    try {
+                        cls = (exprType) popNode;
+                    } catch (Exception e) {
+                        Log.log("Expected expr. Found: " + popNode);
+                    }
+                    return new MatchClass(cls, patterns, kwd_attrs, kwd_patterns);
+                } else if (arity == 1) {
+                    return stack.popNode();
```

Comment:
This seems a bit strange... I think that you should change to always create the `MatchClass` here and make the `attr` (to be renamed to `match_class_attr`) return the name which will be processed.